{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a3ea14ec-293d-48c5-8d2a-ac4ecb5cd4a8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#! conda install matplotlib torch pandas numpy scikit-learn torchvision tensorflow tensorboard tqdm seaborn biopython xgboost \n",
    "#! pip install jupyter-server-proxy viennaRNA torchsummary "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "555d58d4-777b-4bb0-b238-187f37ba4a2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-08-12 18:22:47.717065: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.\n",
      "2025-08-12 18:22:48.790965: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: SSE4.1 SSE4.2 AVX AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    }
   ],
   "source": [
    "import datetime\n",
    "import itertools\n",
    "from typing import Literal, Optional\n",
    "\n",
    "import Bio.Seq\n",
    "from Bio import SeqIO\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import sklearn\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import accuracy_score, average_precision_score, roc_auc_score, r2_score\n",
    "from sklearn.preprocessing import LabelEncoder, OneHotEncoder\n",
    "import tensorflow as tf\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.optim as optim\n",
    "from torch.utils.data import DataLoader, TensorDataset, random_split\n",
    "from torch.utils.tensorboard import SummaryWriter\n",
    "from torchvision import datasets\n",
    "from torchvision.transforms import ToTensor\n",
    "from torchsummary import summary\n",
    "from torchviz import make_dot\n",
    "import tqdm\n",
    "from xgboost import XGBRegressor, XGBClassifier\n",
    "import ViennaRNA\n",
    "\n",
    "import data_loading\n",
    "import models\n",
    "import train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6321eaf6-aa2b-49b1-b812-c39a842768ac",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%load_ext tensorboard\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e98991b8-9adf-49bf-8733-414016892b83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gfp_minus_rate</th>\n",
       "      <th>gfp_plus_rate</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>prf_class</th>\n",
       "      <th>gfp_minus</th>\n",
       "      <th>gfp_plus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1506</th>\n",
       "      <td>0.86</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GATAGGTACCTGGCTGTCGTCCATGCTGTGTTTGCAAAAAAAGCCA...</td>\n",
       "      <td>37296</td>\n",
       "      <td>human - ccr5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1507</th>\n",
       "      <td>58.57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GATAGGTACCTGGCTGTCGTCCATGCTGTGTTTGCAAAAAACGCCA...</td>\n",
       "      <td>37297</td>\n",
       "      <td>human - ccr5</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1508</th>\n",
       "      <td>NaN</td>\n",
       "      <td>71.48</td>\n",
       "      <td>GATAGGTACCTGGCTGTCGTCCATGCTGTGTTTGCAAAAAAGGCCA...</td>\n",
       "      <td>37298</td>\n",
       "      <td>human - ccr5</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1509</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.30</td>\n",
       "      <td>GATAGGTACCTGGCTGTCGTCCATGCTGTGTTTGCAAAAAATGCCA...</td>\n",
       "      <td>37299</td>\n",
       "      <td>human - ccr5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1511</th>\n",
       "      <td>NaN</td>\n",
       "      <td>47.30</td>\n",
       "      <td>GATAGGTACCTGGCTGTCGTCCATGCTGTGTTTGCAAACCCCGCCA...</td>\n",
       "      <td>37301</td>\n",
       "      <td>human - ccr5</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gfp_minus_rate  gfp_plus_rate  \\\n",
       "1506            0.86            NaN   \n",
       "1507           58.57            NaN   \n",
       "1508             NaN          71.48   \n",
       "1509             NaN           1.30   \n",
       "1511             NaN          47.30   \n",
       "\n",
       "                                               sequence  variant_id  \\\n",
       "1506  GATAGGTACCTGGCTGTCGTCCATGCTGTGTTTGCAAAAAAAGCCA...       37296   \n",
       "1507  GATAGGTACCTGGCTGTCGTCCATGCTGTGTTTGCAAAAAACGCCA...       37297   \n",
       "1508  GATAGGTACCTGGCTGTCGTCCATGCTGTGTTTGCAAAAAAGGCCA...       37298   \n",
       "1509  GATAGGTACCTGGCTGTCGTCCATGCTGTGTTTGCAAAAAATGCCA...       37299   \n",
       "1511  GATAGGTACCTGGCTGTCGTCCATGCTGTGTTTGCAAACCCCGCCA...       37301   \n",
       "\n",
       "         prf_class  gfp_minus  gfp_plus  \n",
       "1506  human - ccr5      False     False  \n",
       "1507  human - ccr5       True     False  \n",
       "1508  human - ccr5      False      True  \n",
       "1509  human - ccr5      False     False  \n",
       "1511  human - ccr5      False      True  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = data_loading.build_dataset('frameshift_data_natcomm.csv') #supplementary data 2 file of paper\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0c391113-ad6d-4e69-88e5-e9da67724d06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "prf_class\n",
       "HIV HXB2                             614\n",
       "PLRV luteovirus                      612\n",
       "human - OAZ1                         596\n",
       "SARS coronavirus                     583\n",
       "herpes simplex - thymidine kinase    582\n",
       "human - PEG10                        575\n",
       "simian srv1 retrovirus               562\n",
       "human - HERV-K10                     525\n",
       "Rous sarcoma virus                   511\n",
       "influenza a virus                    456\n",
       "SIVmac239                            450\n",
       "synthetic                            450\n",
       "west nile virus                      440\n",
       "PRRSV - nsp2TF                       422\n",
       "HCV - F protein                      378\n",
       "human T-lymphotropic virus           369\n",
       "human - ccr5                         276\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['prf_class'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "22ebdc5e-d44e-4c65-b0d4-8d03afa7624b",
   "metadata": {},
   "outputs": [],
   "source": [
    "PAPER_CAT = [\n",
    "    'SARS coronavirus',\n",
    "    'SIVmac239',\n",
    "    'simian srv1 retrovirus',\n",
    "    'PLRV luteovirus',\n",
    "    'human - HERV-K10',\n",
    "    'human T-lymphotropic virus',\n",
    "    'HIV HXB2',\n",
    "    'west nile virus',\n",
    "    'mix'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "25a09653-c075-46cb-b474-22c188a0077f",
   "metadata": {},
   "outputs": [],
   "source": [
    "baselines = {\n",
    "    'mix': {\n",
    "        'ROC-AUC/val': .89,\n",
    "        'ROC-AUC/train': .89,\n",
    "        'PR-AUC/val': .8,\n",
    "        'PR-AUC/train': .8,\n",
    "        'Pearson/val': .54,\n",
    "        'Pearson/train': .54,\n",
    "    },\n",
    "    'SARS coronavirus': {\n",
    "        'ROC-AUC/val': .63,\n",
    "        'ROC-AUC/train': .63, \n",
    "        'Pearson/val': .24,\n",
    "        'Pearson/train': .24,\n",
    "    },\n",
    "    'SIVmac239': {\n",
    "        'ROC-AUC/val': .74,\n",
    "        'ROC-AUC/train': .74,  \n",
    "        'Pearson/val': .45,\n",
    "        'Pearson/train': .45,\n",
    "    },\n",
    "    'simian srv1 retrovirus': {\n",
    "        'ROC-AUC/val': .94,\n",
    "        'ROC-AUC/train': .94,    \n",
    "        'Pearson/val': .61,\n",
    "        'Pearson/train': .61,\n",
    "    },\n",
    "    'PLRV luteovirus': {\n",
    "        'ROC-AUC/val': .66,\n",
    "        'ROC-AUC/train': .66,    \n",
    "        'Pearson/val': .19,\n",
    "        'Pearson/train': .19,\n",
    "    },\n",
    "    'human - HERV-K10': {\n",
    "        'ROC-AUC/val': .85,\n",
    "        'ROC-AUC/train': .85,   \n",
    "        'Pearson/val': .63,\n",
    "        'Pearson/train': .63,\n",
    "    },\n",
    "    'human T-lymphotropic virus': {\n",
    "        'ROC-AUC/val': .88,\n",
    "        'ROC-AUC/train': .88,   \n",
    "        'Pearson/val': .39,\n",
    "        'Pearson/train': .39,\n",
    "    },\n",
    "    'HIV HXB2': {\n",
    "        'ROC-AUC/val': .89,\n",
    "        'ROC-AUC/train': .89,      \n",
    "        'Pearson/val': .69,\n",
    "        'Pearson/train': .69,\n",
    "    },\n",
    "    'west nile virus': {\n",
    "        'ROC-AUC/val': .47,\n",
    "        'ROC-AUC/train': .47,   \n",
    "        'Pearson/val': .61,\n",
    "        'Pearson/train': .61,\n",
    "    },\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "raw",
   "id": "2b124655-8f2b-4cfb-8e09-746c5c29eff8",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "Variant ID\tSequence\n",
    "48119\tSARS\n",
    "48120\tSARS\n",
    "48121\tSARS\n",
    "48146\tHERV\n",
    "48147\tHERV\n",
    "48148\tHERV\n",
    "48167\tHIV\n",
    "48168\tHIV\n",
    "48169\tHIV\n",
    "48152\tHTLV\n",
    "48153\tHTLV\n",
    "48154\tHTLV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f85bc900-3964-4d7a-b5e6-19a8bc7dcf2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def secondary_features(sequence: str):\n",
    "    _MAPPING = {'.': 0, '(': 1, ')': 1} \n",
    "    folded, _ = ViennaRNA.fold(sequence)\n",
    "    encoded = [_MAPPING[pos] for pos in folded]\n",
    "    return encoded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c0da2709-6712-4d9c-b476-20eb8a44682d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def aa_polarity_features(sequence: str):\n",
    "    _AA_POLARITY = {\n",
    "        'R':'charged',\n",
    "        'H':'charged',\n",
    "        'K':'charged',\n",
    "        'D':'charged',\n",
    "        'E':'charged',\n",
    "        'S':'polar',\n",
    "        'T':'polar',\n",
    "        'N':'polar',\n",
    "        'Q':'polar',\n",
    "        'C':'polar',\n",
    "        'Y':'polar',\n",
    "        'G':'unpolar',\n",
    "        'P':'unpolar',\n",
    "        'A':'unpolar',\n",
    "        'V':'unpolar',\n",
    "        'I':'unpolar',\n",
    "        'L':'unpolar',\n",
    "        'M':'unpolar',\n",
    "        'F':'unpolar',\n",
    "        'W':'unpolar',\n",
    "        '*':'stop'\n",
    "    }\n",
    "    translated = Bio.Seq.Seq(df.iloc[0].sequence).translate()\n",
    "    polarity = [_AA_POLARITY[aa] for aa in translated]\n",
    "    return polarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "eef0f9e6-b651-4810-b0e5-104599a3924a",
   "metadata": {},
   "outputs": [],
   "source": [
    "tai_availability = pd.read_csv('tAI_index_human_nar-02315.txt', sep='\\t').rename(columns={'w ': 'w'})\n",
    "tai_availability.loc[:, 'w'] = tai_availability.loc[:, 'w'].map(lambda x: float(x))\n",
    "tai_dict = dict(zip(tai_availability['Codon'], tai_availability['w']))\n",
    "#tai_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "17fa0fe2-9658-4b7b-a600-8f42b5e044ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mikl_tai(df, tai_file):\n",
    "    tai_availability = pd.read_csv(tai_file, sep='\\t').rename(columns={'w ': 'w'})\n",
    "    tai_availability.loc[:, 'w'] = tai_availability.loc[:, 'w'].map(lambda x: float(x))\n",
    "    tai_dict = dict(zip(tai_availability['Codon'], tai_availability['w']))\n",
    "\n",
    "    def _gettaimap(vrsq):\n",
    "        tai=[]\n",
    "        for i in range(len(vrsq)//3-1):\n",
    "            tai.append(tai_dict[vrsq[0+3*i:3+3*i]])\n",
    "        return pd.Series(tai)\n",
    "    \n",
    "    def _gettaidiffmap(vrsq):\n",
    "        tai=[]\n",
    "        for i in range(len(vrsq)//3-1):\n",
    "            tai.append(tai_dict[vrsq[2+3*i:5+3*i]]-tai_dict[vrsq[3+3*i:6+3*i]])\n",
    "        return pd.Series(tai)\n",
    "\n",
    "    tai0=df.sequence.apply(lambda x: _gettaimap(x.upper()))\n",
    "    taim1=df.sequence.apply(lambda x: _gettaimap(x[2:].upper()))\n",
    "    taip1=df.sequence.apply(lambda x: _gettaimap(x[1:].upper()))\n",
    "    taidiff=df.sequence.apply(lambda x: _gettaidiffmap(x.upper()))\n",
    "\n",
    "    tai=tai0.join(taim1, lsuffix='_frame0', rsuffix = '_framem1')\n",
    "    tai=tai.join(taip1, rsuffix = '_framep1')\n",
    "    tai=tai.join(taidiff, rsuffix = '_diffframem1to0')\n",
    "\n",
    "    return tai\n",
    "\n",
    "def mikl_charged(df):\n",
    "    aacategory={\n",
    "        'R':'charged',\n",
    "        'H':'charged',\n",
    "        'K':'charged',\n",
    "        'D':'charged',\n",
    "        'E':'charged',\n",
    "        'S':'polar',\n",
    "        'T':'polar',\n",
    "        'N':'polar',\n",
    "        'Q':'polar',\n",
    "        'C':'polar',\n",
    "        'Y':'polar',\n",
    "        'G':'unpolar',\n",
    "        'P':'unpolar',\n",
    "        'A':'unpolar',\n",
    "        'V':'unpolar',\n",
    "        'I':'unpolar',\n",
    "        'L':'unpolar',\n",
    "        'M':'unpolar',\n",
    "        'F':'unpolar',\n",
    "        'W':'unpolar',\n",
    "        '*':'stop'\n",
    "        }\n",
    "    \n",
    "    def getaaclassmap(vrsq):\n",
    "        aa=[]\n",
    "        for i in range(len(vrsq)//3-1):\n",
    "            aa.append(aacategory[Bio.Seq.Seq(vrsq[0+3*i:3+3*i]).translate()])\n",
    "        return pd.Series(aa)\n",
    "    \n",
    "    aa0=df.sequence.apply(lambda x: getaaclassmap(x.upper()))\n",
    "    aam1=df.sequence.apply(lambda x: getaaclassmap(x[2:].upper()))\n",
    "    aap1=df.sequence.apply(lambda x: getaaclassmap(x[1:].upper()))\n",
    "    \n",
    "    aa0enc=pd.DataFrame(index=aa0.index)\n",
    "    for col in aa0.columns:\n",
    "        labelenc=LabelEncoder()\n",
    "        feature0=labelenc.fit_transform(aa0[col])\n",
    "        onehotenc=OneHotEncoder(sparse_output=False)\n",
    "        aa0enc=aa0enc.join(pd.DataFrame(onehotenc.fit_transform(feature0.reshape(feature0.shape[0],1)), index=aa0.index, columns=labelenc.classes_), rsuffix=col)\n",
    "    \n",
    "    aam1enc=pd.DataFrame(index=aam1.index)\n",
    "    for col in aam1.columns:\n",
    "        labelenc=LabelEncoder()\n",
    "        feature0=labelenc.fit_transform(aam1[col])\n",
    "        onehotenc=OneHotEncoder(sparse_output=False)\n",
    "        aam1enc=aam1enc.join(pd.DataFrame(onehotenc.fit_transform(feature0.reshape(feature0.shape[0],1)), index=aam1.index, columns=labelenc.classes_), rsuffix=col)\n",
    "    \n",
    "    aap1enc=pd.DataFrame(index=aap1.index)\n",
    "    for col in aap1.columns:\n",
    "        labelenc=LabelEncoder()\n",
    "        feature0=labelenc.fit_transform(aap1[col])\n",
    "        onehotenc=OneHotEncoder(sparse_output=False)\n",
    "        aap1enc=aap1enc.join(pd.DataFrame(onehotenc.fit_transform(feature0.reshape(feature0.shape[0],1)), index=aap1.index, columns=labelenc.classes_), rsuffix=col)\n",
    "    \n",
    "    aa=aa0enc.join(aam1enc, lsuffix='_frame0', rsuffix = '_framem1')\n",
    "    aa=aa.join(aap1enc, rsuffix = '_framep1')\n",
    "    return aa\n",
    "\n",
    "def mikl_fold(df):\n",
    "    sec=df.sequence.apply(lambda x: pd.Series(list(ViennaRNA.fold(x)[0])))\n",
    "    sec=sec.replace({'.':0,'(':1,')':1})\n",
    "    return sec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d8dd5c8-27f0-4a2c-a257-7e9a8d75fe60",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "aa = mikl_charged(df)\n",
    "sec = mikl_fold(df)\n",
    "tai = mikl_tai(df, 'tAI_index_human_nar-02315.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "739db57b-4de0-4e93-9dea-84ee007687f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "feat=tai.join(aa)\n",
    "#feat=feat.join(fsdesforml[[x for x in fsdesforml.columns if 'dg' in x]])\n",
    "feat=feat.join(sec)\n",
    "feat=feat.join(df)\n",
    "#feat=feat.join(slipenc)\n",
    "seccols=list(np.arange(42,162))\n",
    "taicols=[str(x)+'_frame0' for x in np.arange(12,26)]+[str(x)+'_framem1' for x in np.arange(12,26)]+[str(x)+'_diffframem1to0' for x in np.arange(12,26)]\n",
    "aacols=['charged'+str(x) for x in np.arange(10, 14)]+['polar'+str(x) for x in np.arange(10, 14)]+['unpolar'+str(x) for x in np.arange(10, 14)]\n",
    "all_cols = seccols+taicols+aacols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee6e4e2c-abb5-46fb-9b5c-a1ffc2a6552d",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_splits=list(itertools.product(PAPER_CAT, PAPER_CAT))\n",
    "\n",
    "mikl_perfs = []\n",
    "\n",
    "for x, y in dataset_splits:\n",
    "    if x == 'mix' or y == 'mix':\n",
    "        continue\n",
    "    if x == y:\n",
    "        train_data, test_data = train_test_split(feat[feat['prf_class'] == x], test_size=.2, random_state=42)\n",
    "    else:\n",
    "        train_data = feat[feat['prf_class'] == x]\n",
    "        test_data = feat[feat['prf_class'] == y]\n",
    "    model = XGBClassifier()\n",
    "    model.fit(train_data[all_cols], train_data['gfp_minus'])\n",
    "    pred = model.predict_proba(test_data[all_cols])\n",
    "    roc = roc_auc_score(test_data['gfp_minus'], pred[:,1])\n",
    "    mikl_perfs.append({\n",
    "        'train_data': x,\n",
    "        'test_data': y,\n",
    "        'roc_auc': roc\n",
    "    })\n",
    "mikl_df = pd.DataFrame.from_records(mikl_perfs)\n",
    "tmp = mikl_df.pivot_table(columns=\"test_data\", index=\"train_data\", values='roc_auc', aggfunc=\"mean\")\n",
    "tmp = tmp.reindex(index=PAPER_CAT, columns=PAPER_CAT)\n",
    "sns.heatmap(tmp, annot=True, fmt=\".3f\", linewidths=0.5, vmin=0, vmax=0.82, cmap=sns.cubehelix_palette(as_cmap=True))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d7ba260-5cb6-4b63-887f-a25e6e5d3148",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikl_perfs = []\n",
    "\n",
    "ablation = {\n",
    "    'all': all_cols,\n",
    "    'no tRNA availability': seccols+aacols,\n",
    "    'no RNA secondary structure': taicols+aacols,\n",
    "    'no AA charge': seccols+taicols,\n",
    "    'only AA charge': aacols,\n",
    "    'only tRNA availability': taicols,\n",
    "    'only RNA secondary structure': seccols\n",
    "}\n",
    "\n",
    "for method in ablation:\n",
    "    for x in PAPER_CAT:\n",
    "        if x == 'mix':\n",
    "            train_data, test_data = train_test_split(feat, test_size=.2, random_state=42)\n",
    "        else:\n",
    "            train_data = feat[feat['prf_class'] != x]\n",
    "            test_data = feat[feat['prf_class'] == x]\n",
    "        model = XGBClassifier()\n",
    "        model.fit(train_data[ablation[method]], train_data['gfp_minus'])\n",
    "        pred = model.predict_proba(test_data[ablation[method]])\n",
    "        mikl_perfs.append({\n",
    "            'train_data': 'leave_one',\n",
    "            'test_data': x,\n",
    "            'roc_auc': roc_auc_score(test_data['gfp_minus'], pred[:,1]),\n",
    "            'pr_auc': average_precision_score(test_data['gfp_minus'], pred[:,1]),\n",
    "            'model': f'Mikl {method}'\n",
    "        })\n",
    "\n",
    "mikl_df = pd.DataFrame.from_records(mikl_perfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af49ba0c-f969-43a1-9e6c-fbbf838800fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikl_perfs = []\n",
    "\n",
    "ablation = {\n",
    "    'all': all_cols,\n",
    "    'no tRNA availability': seccols+aacols,\n",
    "    'no RNA secondary structure': taicols+aacols,\n",
    "    'no AA charge': seccols+taicols,\n",
    "    'only AA charge': aacols,\n",
    "    'only tRNA availability': taicols,\n",
    "    'only RNA secondary structure': seccols\n",
    "}\n",
    "\n",
    "for i in range(10):\n",
    "    for method in ablation:\n",
    "        for x in PAPER_CAT:\n",
    "            if x == 'mix':\n",
    "                train_data, test_data = train_test_split(feat, test_size=.2, random_state=42)\n",
    "            else:\n",
    "                train_data = feat[feat['prf_class'] != x]\n",
    "                test_data = feat[feat['prf_class'] == x]\n",
    "            model = XGBClassifier(seed = i)\n",
    "            model.fit(train_data[ablation[method]], train_data['gfp_minus'])\n",
    "            pred = model.predict_proba(test_data[ablation[method]])\n",
    "            mikl_perfs.append({\n",
    "                'train_data': 'leave_one',\n",
    "                'test_data': x,\n",
    "                'roc_auc': roc_auc_score(test_data['gfp_minus'], pred[:,1]),\n",
    "                'pr_auc': average_precision_score(test_data['gfp_minus'], pred[:,1]),\n",
    "                'model': f'Mikl {method}',\n",
    "                'iter': i\n",
    "            })\n",
    "\n",
    "mikl_df = pd.DataFrame.from_records(mikl_perfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f031947a-493a-4920-8a92-2d125e735ca8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = sns.barplot(mikl_df, x='roc_auc', y='test_data', hue='model')\n",
    "sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "plt.xlim(xmin=0.4)\n",
    "plt.show()\n",
    "\n",
    "ax = sns.barplot(mikl_df, x='pr_auc', y='test_data', hue='model')\n",
    "sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "plt.xlim(xmin=0.4)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db2fcf3e-b418-496c-bbdc-760b65e1b3c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Instantiate the model\n",
    "model = models.RNACNN_pete()\n",
    "\n",
    "input_shape = (5, 162)\n",
    "\n",
    "# Generate summary\n",
    "summary(model, input_shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0de85747-db7d-470f-9c61-db4136579f9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "\n",
    "cnn_df = train.main_loop(\n",
    "    model_fns=[\n",
    "        lambda: models.RNACNN_pete()\n",
    "        #lambda: models.RNACNN_pete()\n",
    "        #lambda: models.RNACNN(),\n",
    "        #lambda: models.RNACNN()\n",
    "        #lambda: models.RNACNN_FC(),\n",
    "        #lambda: models.RNACNN_GN(),\n",
    "        #lambda: models.RNACNN_FC_GN(),\n",
    "     ],\n",
    "    data=df,\n",
    "    device=device,\n",
    "    base_name='cnn',\n",
    "    lrs=[1e-3,],\n",
    "    #lrs=[1e-2, 1e-3, 1e-4, 1e-5,],\n",
    "    wds=[0.01],\n",
    "    #wds=[0.01, 0.05, 0.1, 0.2],\n",
    "    schedulers = ['CyclicLR',],\n",
    "    #schedulers = ['StepLR', 'ExponentialLR', 'CosineAnnealingLR', 'CyclicLR'],\n",
    "    dataset_splits=PAPER_CAT,\n",
    "    #baselines=baselines,\n",
    "    epochs=200,\n",
    "    loss_mixin=.999,\n",
    "    data_augment=False\n",
    ")\n",
    "cnn_df = cnn_df.rename(columns={'mix': 'test_data'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17d0333b-bca9-4e67-b531-590026babff2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from typing import Mapping\n",
    "\n",
    "def load_trna_mapping(file_path: str) -> Mapping[str, float]:\n",
    "    df = pd.read_csv(file_path, sep='\\t')\n",
    "    return dict(zip(df['Codon'], df['w']))\n",
    "\n",
    "trna_av = data_loading.load_trna_mapping(\"tAI_index_human_nar-02315.txt\")\n",
    "print(trna_av[\"ATG\"])  # Should print 0.611247 (w value for AUG)\n",
    "print(trna_av[\"TTT\"])  # Should print 0.216381"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6662d86-6ef0-4018-b763-e9091781aaff",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "print(f\"Device: {device}\")\n",
    "print(f\"GPU available: {torch.cuda.is_available()}\")\n",
    "print(f\"GPU count: {torch.cuda.device_count()}\")\n",
    "print(f\"GPU name: {torch.cuda.get_device_name(0) if torch.cuda.is_available() else 'None'}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98ae3ce7-7e29-47f2-b6b0-0fc76a6faf23",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnn_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18bf1b9a-2eb2-4f1e-bbc7-184883317aa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_baseline = []\n",
    "for x in PAPER_CAT:\n",
    "    if x == 'mix':\n",
    "        train_data, test_data = train_test_split(feat, test_size=.2, random_state=42)\n",
    "    else:\n",
    "        train_data = feat[feat['prf_class'] != x]\n",
    "        test_data = feat[feat['prf_class'] == x]\n",
    "    random_baseline.append({\n",
    "        'test_data': x,\n",
    "        'roc_auc': .5,\n",
    "        'pr_auc': test_data['gfp_minus'].mean(),\n",
    "        'model': f'Random baseline'\n",
    "    })\n",
    "\n",
    "total = pd.concat([cnn_df, mikl_df, pd.DataFrame.from_records(random_baseline)])\n",
    "\n",
    "\n",
    "tmp = total.pivot_table(index=\"test_data\", columns=\"model\", values='roc_auc', aggfunc=\"mean\")\n",
    "tmp = tmp.reindex(columns=sorted(total['model'].unique()), index=PAPER_CAT)\n",
    "sns.heatmap(tmp, annot=True, fmt=\".3f\", linewidths=0.5, cmap=sns.cubehelix_palette(as_cmap=True))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f1afe84-c208-4e0c-ada5-451fea3c0dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "mikl_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02258dcb-b426-4b93-b017-78b3bb71b585",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = {\n",
    "    'all': all_cols,\n",
    "    'no tRNA availability': seccols+aacols,\n",
    "    'no RNA secondary structure': taicols+aacols,\n",
    "    'no AA charge': seccols+taicols,\n",
    "    'only AA charge': aacols,\n",
    "    'only tRNA availability': taicols,\n",
    "    'only RNA secondary structure': seccols\n",
    "}\n",
    "\n",
    "MIKL_ORDER = [\n",
    "    'Mikl all',\n",
    "    'Mikl only tRNA availability',\n",
    "    'Mikl no tRNA availability'\n",
    "]\n",
    "\n",
    "ax = sns.barplot(mikl_df, x='roc_auc', y='test_data', hue='model', hue_order=MIKL_ORDER)\n",
    "sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "plt.xlim(xmin=0.4)\n",
    "plt.show()\n",
    "\n",
    "ax = sns.barplot(mikl_df, x='pr_auc', y='test_data', hue='model', hue_order=MIKL_ORDER)\n",
    "sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "plt.xlim(xmin=0.4)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40c19b06-683c-43f5-8dc5-ed5512db97f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = sns.barplot(total, x='roc_auc', y='test_data', hue='model', hue_order=['RNACNN_pete', 'Mikl all', 'Random baseline'])\n",
    "sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "plt.xlim(xmin=0.4)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7f3524a-c2a8-4f80-80f5-039c74b11620",
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = sns.barplot(total, x='roc_auc', y='test_data', hue='model', hue_order=['RNACNN_pete', 'Mikl all', 'Random baseline'])\n",
    "sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "plt.xlim(xmin=0.4)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7d3a578-a7ba-44bd-a5ab-6bb6903eb344",
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = sns.barplot(total, x='pr_auc', y='test_data', hue='model', hue_order=['RNACNN_pete', 'Mikl all', 'Random baseline'])\n",
    "#plt.xlim(xmin=0.5)\n",
    "sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "387babfc-d011-432d-89b8-4a2f9a5d60e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = sns.barplot(total, x='pr_auc', y='test_data', hue='model', hue_order=['RNACNN_pete', 'Mikl all', 'Random baseline'])\n",
    "#plt.xlim(xmin=0.5)\n",
    "sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbcc475a-23c8-44da-9527-5bfaa1499fe8",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20cc34dd-ef8c-4ea7-aceb-0a423fc3e6b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "total[['roc_auc', 'model']].groupby('model').mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9bdaae2-03f5-406c-a674-0cb23e65e08f",
   "metadata": {},
   "outputs": [],
   "source": [
    "total[['roc_auc', 'model']].groupby('model').mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59738f43-b8f6-4155-8347-f05342081e61",
   "metadata": {},
   "outputs": [],
   "source": [
    "total[['pr_auc', 'model']].groupby('model').mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f6ad9b1-3396-4749-be23-3d952f82580c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import collections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f861860-bab8-4f9d-bf64-3cb833614abb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14eaf76c-dae5-428f-9070-4d8acdc32c06",
   "metadata": {},
   "outputs": [],
   "source": [
    "acc = []\n",
    "for m in sorted(total['model'].unique()):\n",
    "    acc.append({\n",
    "        'Method': m,\n",
    "        'Mean AURoC': total[total['model'] == m]['roc_auc'].mean(),\n",
    "        'Standard Deviation AURoC': total[total['model'] == m]['roc_auc'].std(),\n",
    "        'Mean AUPR': total[total['model'] == m]['pr_auc'].mean(),\n",
    "        'Standard Deviation AUPR': total[total['model'] == m]['pr_auc'].std(),\n",
    "    })\n",
    "total_perfs = pd.DataFrame.from_records(acc)\n",
    "total_perfs.sort_values('Mean AUPR', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61752c6d-8d08-433c-acbd-d82baa129d26",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2da6e4e-c146-45b7-a662-fab712887342",
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluate_sequence(model: nn.Module, transcript: str, window_size: int, start: int = 0, end: Optional[int] = None):\n",
    "    acc = []\n",
    "    if end is None:\n",
    "        end = len(transcript) - window_size\n",
    "        \n",
    "    for i in range(start, end, 3):\n",
    "        acc.append({\n",
    "            'sequence': str(transcript[i:i+window_size]),\n",
    "            'start': i,\n",
    "            'end': i+window_size,\n",
    "            'peak': int(i + (window_size/2))\n",
    "        })\n",
    "    return acc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "808a7dbd-74fd-4175-ac97-4166c6b53801",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.iloc[0].sequence[30:50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45792661-5120-476d-9636-afd2cbea2e16",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efcf2094-ce74-4a9a-9835-b24073394b63",
   "metadata": {},
   "outputs": [],
   "source": [
    "torch.save(model.state_dict(), 'rnacnn.pth')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
